pulsar-client-cpp/lib/ClientImpl.cc

```diff
@@ -536,8 +536,13 @@ void ClientImpl::handleClose(Result result, SharedInt numberOfOpenHandlers, Resu
     }
     if (*numberOfOpenHandlers == 0) {
         Lock lock(mutex_);
-        state_ = Closed;
-        lock.unlock();
+        if (state_ == Closed) {
+            LOG_DEBUG("Client is already shutting down, possible race condition in handleClose");
+            return lock.unlock();
```

Comment:
```suggestion
            return;
```
The `unlock` method could be invoked automatically.